### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by unfalse**
Task: 4

addProduct: https://xrmedhz5nf.execute-api.eu-west-1.amazonaws.com/dev/products
getProductId: https://xrmedhz5nf.execute-api.eu-west-1.amazonaws.com/dev/products/%7BproductId%7D
getProductsList: https://xrmedhz5nf.execute-api.eu-west-1.amazonaws.com/dev/products

FE: https://d2p1ss31jpj477.cloudfront.net/

### CODE REVIEW

**Comment by ReshetovItsMe** [comment]
ping me when you will be ready please

**Comment by ReshetovItsMe** [comment]
approve

File: product-service/src/functions/addProduct/handler.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+export const addProduct: APIGatewayProxyHandler = async (event) => {
+    console.log('addProduct lambda is executing', {
+        method: event.httpMethod,
+        pathParameters: event.pathParameters,
+        body: event.body
+    });
+
+    try {
+        const product = (event.body as Object) as Product;
+        product.id = v4();
```
Comment (AndreyAusianik) [inline product-service/src/functions/addProduct/handler.ts]: generating id inside of server code seems strange to me

File: product-service/src/functions/addProduct/handler.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        type: "body",
+        parameter: "title",
+        errorMessage: "title is a required field",
+        validationFunction: (title: string) => typeof title === "string" && title.length > 0,
+    },
+    {
+        type: "body",
+        parameter: "price",
+        errorMessage: "price must be non negative integer",
+        validationFunction: (price: number) => typeof price === 'number' && Number.isInteger(price) && price >= 0,
```
Comment (AndreyAusianik) [inline product-service/src/functions/addProduct/handler.ts:46]: I believe there is a builtin validation that does exactly the same checks, but idk if it is the best practice

File: product-service/src/functions/getProductById/handler.ts [hunk 3]
```diff
@@ -4,24 +4,28 @@ import { APIGatewayProxyHandler } from 'aws-lambda';
...
             return formatJSONResponse(product);
         }
         return formatJSONResponse({
             body: 'Product not found'
         }, STATUS_CODES.NOT_FOUND);
     } catch (error) {
-        return formatJSONResponse({
-            body: `Something bad has happened: ${JSON.stringify(error)}`
-        }, STATUS_CODES.BAD_REQUEST);
+        throw error;
```
Comment (AndreyAusianik) [inline product-service/src/functions/getProductById/handler.ts]: don't really see a reason to just catch and rethrow an eror. did you forget to add logging here?

File: product-service/src/functions/getProductsList/handler.ts [hunk 4]
```diff
@@ -4,19 +4,20 @@ import { APIGatewayProxyHandler } from 'aws-lambda';
...
-        return formatJSONResponse(
-            await service.getProductsList()
-        );
+        const products = await getProductsListFromDb();
+        return formatJSONResponse(products);
     } catch (error) {
-        return formatJSONResponse({
-            body: `Something bad has happened: ${JSON.stringify(error)}`
-        }, STATUS_CODES.SERVER_ERROR);
+        throw error;
```
Comment (AndreyAusianik) [inline product-service/src/functions/getProductsList/handler.ts]: same as above

File: product-service/src/db-scripts/create_products_table.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE Products (
+    id UUID primary key,
```
Comment (AndreyAusianik) [inline product-service/src/db-scripts/create_products_table.sql]: it is possible to make id autogeneratable on database

File: product-service/src/services/db/queries.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+export const getProductById = async id => {
+    const client = await connect();
+    try {
+        const { rows: products } = await client.query(`
+            select id, title, description, price, img_url, s.count 
+            from public.products p left join public.stocks s
+            on p.id = s.product_id
+            where p.id = $1
+        `, [id]);
+        return products;
```
Comment (AndreyAusianik) [inline product-service/src/services/db/queries.ts:28]: it seems here it should be error 404, in case no product is found

**Review by AndreyAusianik — APPROVED** [review]
task is completed, with no reasons to lower the final mark, but I left a few comments that seem reasonable to me

File: product-service/src/functions/getProductById/handler.ts (hunk 3 above)
Comment (unfalse) [reply product-service/src/functions/getProductById/handler.ts]: there is an onError middleware added by middyfy that catch such errors

Comment (AndreyAusianik) [reply product-service/src/functions/getProductById/handler.ts]: I mean that if you weren't try-catching the error at all, the result would be the same

File: product-service/src/functions/addProduct/handler.ts (hunk 1 above)
Comment (unfalse) [reply product-service/src/functions/addProduct/handler.ts]: moved into db

File: product-service/src/services/db/queries.ts (hunk 6 above)
Comment (unfalse) [reply product-service/src/services/db/queries.ts:28]: it is already in handler

**Pull request merged**